added function returnIfPossible, to leave certificate error

njtwiseman opened the pull request:
Somewhat better:

Added a function "returnIfPossible" which

A. Checks to see if the browser is able to go back
B. Goes back if the above returns true.

Problem is that it seems like it goes back twice now.  Just wanted to submit a little halfway point work since this is the first time I'm contributing

njtwiseman committed:
added function returnIfPossible, to leave certificate error

UniversalSuperBox reviewed (changes requested):
This review should block the merge of this until we can figure out the "going back twice" problem.

I wonder if the problem is that you're connecting to `rejectCertificate`. What does the handler for that signal do?

njtwiseman commented:
So just to clarify that we're both on the same page (haha) 

I start on: 
ubports.com > badssl.com > click on the "Expired" button
When I reject the tab goes away a second or two passes and then the browser returns to 
ubports.com < 2 (We want to be here?) <  1

I think `rejectCertificate` sends something into the background of QML / the webView - haven't been able to find out yet.  Documentation says it _must_ be called when you revoke.

It also says, after reading through it a second time, that it Ends the defer() command. Which, at this point, has paused the URL request.  Upon receiving either the `ignoreError` or reject signals The url request is either continued for ignore or dropped for reject.  

Which makes me think the back command is unnecessary. Unless there's more than one request included in the bad certificate request response exchange -> leaving you in no man's land without a place to go and not having a place to be either.

njtwiseman commented:
Well that wasn't so bad

njtwiseman committed:
Deferred the rejectCertificate method so cvD.reject is not called twice

njtwiseman reviewed (commented):
Seems to be working at this time.

I suspect `certificateVerificationDialog.reject` being called twice was the problem.

UniversalSuperBox commented on src/app/WebViewImpl.qml:
```diff
@@ -248,18 +248,32 @@ WebView {
+        certificateVerificationDialog.reject.connect(returnIfPossible)
+        
+        
+    }
```
This many newlines seems like overkill

njtwiseman replied:
yup

UniversalSuperBox reviewed (changes requested):
Only a style change, it seems to function correctly.

njtwiseman committed:
style changes

njtwiseman commented on src/app/WebViewImpl.qml:
```diff
@@ -313,7 +306,7 @@ WebView {
                 }
 
-                var commandGetContextMenuInfo = "
+                var commandGetContextMenuInfo = ""
```
This didn't look intended so I added a close quote

njtwiseman committed:
Merge branch 'xenial' into leave_evil_site

balcy committed:
Update WebViewImpl.qml

avoid regression of https://github.com/ubports/morph-browser/issues/167

njtwiseman commented:
I see the problem from #167, incorporated upstream changes -> should be good

balcy commented on src/app/WebViewImpl.qml:
```diff
@@ -248,35 +248,43 @@ WebView {
+    // returnIfPossible -> rejects bad certificate and returns to last page
+    function returnIfPossible(certificateError){
+        // deferred rejection so that c.v.Dialog isn't called twice
+        certificateError.rejectCertificate
```
this line shows `Type Error: Cannot read property 'rejectCertificate' of undefined` in the logfile.
Since QtCreator was showing a warning, I changed it to `certificateError.rejectCertificate()`
But then it was still not working: `TypeError: Cannot call method 'rejectCertificate' of undefined`
so in fact "certificateError" is not passed to the function "returnIfPossible", and does nothing

balcy replied:
... but before putting more effort into it, we should discuss with @UniversalSuperBox and @cibersheep  if maybe a Cancel button is enough, and we could instead add the back button to the error sheet later. (https://github.com/ubports/morph-browser/issues/116)

njtwiseman replied:
So for some reason `certificateError` isn't getting passed to the function.

balcy replied:
I think only arguments passed to the reject() signal of the dialog would get passed to the function

balcy committed:
Update WebViewImpl.qml

"commandGetContextMenuInfo" is executed as JavaScript command (several lines) in the browser

balcy commented:
instead of automatically navigating back from the certificate error dialog, the "back to safety" button is put to the ErrorSheet instead (see comments in the issue)
